### PR TITLE
Add static docs

### PR DESCRIPTION
**Pull request by arnaudmiribel**
# Summary

Solves https://github.com/arnaudmiribel/streamlit-extras/issues/173

Our gallery is down sometimes, it'd be great to have a static docs backup at least, or maybe even turn that into the default.
This PR introduces new static docs [[preview]](https://deploy-preview-182--phenomenal-crepe-0effec.netlify.app/extras/colored_header/) based off [material-mkdocs](https://squidfunk.github.io/mkdocs-material)

image
image

# Content
1. General config for mkdocs-material. Includes a set of plugins to easily annotate docstrings or highlight, color code blocks
2. A GitHub action to make it possible to publish and update docs after every PR
3. A workflow to publish preview docs into Netlify
4. A few edits of our extras docstrings so they can be parsed properly with mkdocstrings
5. Embedded stlite iframes 
6. Auto-resize of those iframes

# How does mkdocs work?

## Locally

1. Install docs requirements
```
pip install -r docs/requirements.txt
```

2. Now serve the docs! It should load in http://127.0.0.1:8000
```
mkdocs serve
```

## Publishing

With the following command (use with caution), the docs will get published at https://arnaudmiribel.github.io/streamlit-extras/ - automatically building in the `gh-deploy` branch and reading from there, which is the default GitHub pages looks into.

```
mkdocs gh-deploy --force
```

### CODE REVIEW

File: docs/generate.py
Comment (arnaudmiribel) [inline docs/generate.py:1]: This is an extremely satisfying feature which enables to build static files on the fly, using a Python script.

File: docs/generate.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+import mkdocs_gen_files
+
+import streamlit_extras
+
+extra_modules_names = [
+    extra.name
+    for extra in pkgutil.iter_modules(streamlit_extras.__path__)
+    if extra.name != "query_string"
+]
```
Comment (arnaudmiribel) [inline docs/generate.py:16]: Ah. I excluded that one for some dependency issue - adding a todo that this should be explored more. 
```suggestion
# TODO: Check why query_string fails
extra_modules_names = [
    extra.name
    for extra in pkgutil.iter_modules(streamlit_extras.__path__)
    if extra.name != "query_string"
]
```

File: docs/generate.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+## Docstring
+Visit the [PyPI page](https://pypi.org/project/{pypi_name}/) for more information.
+"""
+
+
+import ast
+import importlib
+
+
+def find_decorated_functions(module_name: str, decorator_name: str = "extra") -> List[str]:
```
Comment (arnaudmiribel) [inline docs/generate.py:73]: We only show docstrings for `@extra` functions and examples.

File: docs/generate.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    for node in ast.walk(parsed_module):
+        if isinstance(node, ast.FunctionDef):
+            for decorator in node.decorator_list:
+                if isinstance(decorator, ast.Name) and decorator.id == decorator_name:
+                    decorated_functions.append(node.name)
+
+    return decorated_functions
+
+
+def find_example_functions(module_name: str) -> List[str]:
```
Comment (blackary) [inline docs/generate.py:103]: Why not just use `__examples__`?

Comment (arnaudmiribel) [reply docs/generate.py]: 100%. Fixed in [f3be42f](https://github.com/arnaudmiribel/streamlit-extras/pull/182/commits/f3be42fd079714afe83e12e1ae058b4a02e0156d) - thanks!

File: .github/workflows/publish_docs.yml [hunk 4]
```diff
@@ -0,0 +1,25 @@
+name: publish_docs
+on:
+  push:
+    branches:
+      - master
```
Comment (blackary) [inline .github/workflows/publish_docs.yml:5]:
```suggestion
```

File: mkdocs.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  - pymdownx.emoji:
+      emoji_index: !!python/name:materialx.emoji.twemoji
+      emoji_generator: !!python/name:materialx.emoji.to_svg
+  - pymdownx.highlight:
+      anchor_linenums: true
+      line_spans: __span
+      pygments_lang_class: true
+  - pymdownx.inlinehilite
+  - pymdownx.snippets
+  - pymdownx.superfences
```
Comment (blackary) [inline mkdocs.yml:59]: Would be good to have a comment about what each of these is used for

File: overrides/main.html [hunk 6]
```diff
@@ -0,0 +1,7 @@
+{% extends "base.html" %}
+
+{% block announce %}
+  <div id="announce">
+    <p>We just released 💠 Grid, 🟰 Row, and more extras! Check out more in this <strong><a style="color:white" href="https://twitter.com/streamlit/status/1689742703657918466">announcement tweet</a></strong> from the Streamlit team.</p>
```
Comment (blackary) [inline overrides/main.html:5]: This is so cool :)

File: src/streamlit_extras/capture/__init__.py [hunk 7]
```diff
@@ -39,21 +39,17 @@ def new_write(b):
...
 @contextmanager
-def stdout(dst: Callable, terminator="\n"):
-    """Capture STDOUT and redirect it to a callable `dst`.
+def stdout(dst: Callable, terminator: str = "\n"):
+    """
+    Capture STDOUT and redirect it to a callable `dst`
 
     Args:
-        dst (callable[str]): A funciton callable with a single string argument. The entire captured contents will be
+        dst (Callable): A funciton callable with a single string argument. The entire captured contents will be
```
Comment (blackary) [inline src/streamlit_extras/capture/__init__.py:47]:
```suggestion
        dst (Callable): A function callable with a single string argument. The entire captured contents will be
```

**Review by blackary — APPROVED** [review]
A few minor comments, overall looks AMAZING!

Could also add stlite = false for camera input live, as the component doesn't seem to work.

I wonder about some day adding an optional "example app" parameter, that would let you directly specify to embed things like https://camera.streamlit.app/ directly 🤷

File: .github/workflows/publish_docs.yml (hunk 4 above)
Comment (sfc-gh-amiribel) [reply .github/workflows/publish_docs.yml:5]: Oops that should be main

File: overrides/main.html (hunk 6 above)
Comment (sfc-gh-amiribel) [reply overrides/main.html:5]: Hey what about st.announce() in Streamlit OSS? 🧠

**Pull request merged**